pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -268,52 +338,151 @@ func (p *DNSProxy) RemoveRestoredRules(endpointID uint16) {
+			delete(allow.nameMatcherBySignature, *entry.key)
+			for k := range entry.mapping {
+				delete(entry.mapping, k)
+			}
+		default:
+			indexEntry.referenceCount -= 1
+		}
+	}
+}
+
+// removeAndReleasePortRulesForID removes the old port rules for the given destPort on the given endpointID. It also
+// releases the matchers so that unused matchers can be freed from memory.
+func (allow perEPAllow) removeAndReleasePortRulesForID(endpointID uint64, destPort uint16) {
+	epPorts, hasEpPorts := allow.allowMap[endpointID]
+	if !hasEpPorts {
+		return
+	}
+	for _, e := range epPorts[destPort] {
+		allow.releasePatternMatcher(e.patternMatcher)
+		allow.releaseNameMatcher(e.nameMatcher)
+	}
+	delete(epPorts, destPort)
+	if len(epPorts) == 0 {
+		delete(allow.allowMap, endpointID)
+	}
+}
+
 // setPortRulesForID sets the matching rules for endpointID and destPort for
-// later lookups. It converts newRules into a unified regexp that can be reused
-// later.
+// later lookups. It converts newRules into a regex for the matchPatterns and a lookup map for the matchNames.
 func (allow perEPAllow) setPortRulesForID(endpointID uint64, destPort uint16, newRules policy.L7DataMap) error {
-	// This is the delete case
 	if len(newRules) == 0 {
-		epPorts := allow[endpointID]
-		delete(epPorts, destPort)
-		if len(epPorts) == 0 {
-			delete(allow, endpointID)
-		}
+		allow.removeAndReleasePortRulesForID(endpointID, destPort)
 		return nil
 	}
+	cse := make(CachedSelectorEntry, len(newRules))
+	var err error
+	for selector, newRuleset := range newRules {
+		regexPattern, fqdnsList := GenerateRegexpAndFqdns(newRuleset)
+
+		regex, err := allow.lookupOrCompilePatternMatcher(regexPattern)
+		if err != nil {
+			break
+		}
 
-	newRE, err := GetSelectorRegexMap(newRules)
+		fqdnMapping := allow.lookupOrCreateNameMatcher(fqdnsList)
+		cse[selector] = fqdnMatcher{
+			patternMatcher: regex,
+			nameMatcher:    fqdnMapping,
+		}
+	}
 	if err != nil {
+		// Unregister all the registered regexps and fqdn mappings before exiting
+		for k, ruleset := range cse {
+			allow.releasePatternMatcher(ruleset.patternMatcher)
+			allow.releaseNameMatcher(ruleset.nameMatcher)
+			delete(cse, k)
+		}
 		return err
-	}
 
-	epPorts, exist := allow[endpointID]
+	}
+	allow.removeAndReleasePortRulesForID(endpointID, destPort)
+	epPorts, exist := allow.allowMap[endpointID]
 	if !exist {
 		epPorts = make(portToSelectorAllow)
-		allow[endpointID] = epPorts
+		allow.allowMap[endpointID] = epPorts
 	}
-
-	epPorts[destPort] = newRE
+	epPorts[destPort] = cse
 	return nil
 }
 
 // setPortRulesForIDFromUnifiedFormat sets the matching rules for endpointID and destPort for
 // later lookups.
-func (allow perEPAllow) setPortRulesForIDFromUnifiedFormat(endpointID uint64, destPort uint16, newRules CachedSelectorREEntry) error {
+func (allow perEPAllow) setPortRulesForIDFromUnifiedFormat(endpointID uint64, destPort uint16, newRules CachedSelectorEntry) error {
+	// TODO this is unused. Should it be removed?
```

Comment:
I think it's called by `(*DNSProxy).UpdateAllowedFromSelectorRegexes()`.

Reply:
Yee, but that one is unused as well it looks like.. 🤔 

Since I have changed the `CachedSelectorEntry` type, I have already altered the signature of that function already in case some other projects depend on it.

Reply:
Yeah that might be the case. Usually for exported functions, it's part of the API. Let's not deal with removing anything now.

Reply:
Hmm. I see, tho. any user of that func after this PR will be unable to utilize the memory benefits of having a separate map for the matchNames, fwiw.


So, in order to be 100% "compatible" it would have to be something like this I guess;

```go
type CachedSelectorREEntry map[policy.CachedSelector]*regexp.Regexp
func (p *DNSProxy) UpdateAllowedFromSelectorRegexes(endpointID uint64, destPort uint16, newRules CachedSelectorREEntry) error {
	p.Lock()
	defer p.Unlock()

	if len(newRules) == 0 {
		allow.removeAndReleasePortRulesForID(endpointID, destPort)
		return nil
	}
	cse := make(CachedSelectorEntry, len(newRules))
	var err error
	for selector, insertedRegex := range newRules {
        // In case the regex is alreay compiled and in use in another matcher, lookupOrInsertPatternMatcher
        // will return a ref. to the existing regex, so use that value instead of the imported one.
		regex := allow.lookupOrInsertPatternMatcher(insertedRegex)

		fqdnMapping := allow.lookup(fqdnsList)
		cse[selector] = fqdnMatcher{
			patternMatcher: regex,
			nameMatcher:    make(map[string]struct{}),
		}
	}

	allow.removeAndReleasePortRulesForID(endpointID, destPort)
	epPorts, exist := allow.allowMap[endpointID]
	if !exist {
		epPorts = make(portToSelectorAllow)
		allow.allowMap[endpointID] = epPorts
	}
	epPorts[destPort] = cse
	return nil
}

func (allow perEPAllow) lookupOrInsertPatternMatcher(regex *regexp.Regexp) (*regexp.Regexp) {
	if entry, ok := allow.patternMatchersByPattern[regex.String()]; ok {
		entry.referenceCount += 1
		return entry.matcher, nil
	}
	allow.patternMatchersByPattern[pattern] = &patternMatcherEntry{matcher: regex, referenceCount: 1}
	return regex
}
```

Reply:
We should add some tests as well then, in case this is considered a stable API that people rely on

Reply:
Looks good, tests would be good as well!